src/libkernel/sscp/amdgpu/atomic.cpp

```diff
@@ -0,0 +1,638 @@
+    __hipsycl_sscp_memory_scope scope, __hipsycl_int8 *ptr, __hipsycl_int8 x) {
+  return __hip_atomic_fetch_or(ptr, x, builtin_memory_order(order),
+                               builtin_memory_scope(scope));
+}
+
+HIPSYCL_SSCP_BUILTIN __hipsycl_int16 __hipsycl_sscp_atomic_fetch_or_i16(
+    __hipsycl_sscp_address_space as, __hipsycl_sscp_memory_order order,
+    __hipsycl_sscp_memory_scope scope, __hipsycl_int16 *ptr, __hipsycl_int16 x) {
+  return __hip_atomic_fetch_or(ptr, x, builtin_memory_order(order),
+                               builtin_memory_scope(scope));
+}
+
+HIPSYCL_SSCP_BUILTIN __hipsycl_int32 __hipsycl_sscp_atomic_fetch_or_i32(
+    __hipsycl_sscp_address_space as, __hipsycl_sscp_memory_order order,
+    __hipsycl_sscp_memory_scope scope, __hipsycl_int32 *ptr, __hipsycl_int32 x) {
+  return __hip_atomic_fetch_or(ptr, x, builtin_memory_order(order),
+                               builtin_memory_scope(scope));
+}
+
+HIPSYCL_SSCP_BUILTIN __hipsycl_int64 __hipsycl_sscp_atomic_fetch_or_i64(
+    __hipsycl_sscp_address_space as, __hipsycl_sscp_memory_order order,
+    __hipsycl_sscp_memory_scope scope, __hipsycl_int64 *ptr, __hipsycl_int64 x) {
+  return __hip_atomic_fetch_or(ptr, x, builtin_memory_order(order),
+                               builtin_memory_scope(scope));
+}
+
+
+
+HIPSYCL_SSCP_BUILTIN __hipsycl_int8 __hipsycl_sscp_atomic_fetch_xor_i8(
+    __hipsycl_sscp_address_space as, __hipsycl_sscp_memory_order order,
+    __hipsycl_sscp_memory_scope scope, __hipsycl_int8 *ptr, __hipsycl_int8 x) {
+  return __hip_atomic_fetch_xor(ptr, x, builtin_memory_order(order),
+                                builtin_memory_scope(scope));
+}
+
+HIPSYCL_SSCP_BUILTIN __hipsycl_int16 __hipsycl_sscp_atomic_fetch_xor_i16(
+    __hipsycl_sscp_address_space as, __hipsycl_sscp_memory_order order,
+    __hipsycl_sscp_memory_scope scope, __hipsycl_int16 *ptr, __hipsycl_int16 x) {
+  return __hip_atomic_fetch_xor(ptr, x, builtin_memory_order(order),
+                                builtin_memory_scope(scope));
+}
+
+HIPSYCL_SSCP_BUILTIN __hipsycl_int32 __hipsycl_sscp_atomic_fetch_xor_i32(
+    __hipsycl_sscp_address_space as, __hipsycl_sscp_memory_order order,
+    __hipsycl_sscp_memory_scope scope, __hipsycl_int32 *ptr, __hipsycl_int32 x) {
+  return __hip_atomic_fetch_xor(ptr, x, builtin_memory_order(order),
+                                builtin_memory_scope(scope));
+}
+
+HIPSYCL_SSCP_BUILTIN __hipsycl_int64 __hipsycl_sscp_atomic_fetch_xor_i64(
+    __hipsycl_sscp_address_space as, __hipsycl_sscp_memory_order order,
+    __hipsycl_sscp_memory_scope scope, __hipsycl_int64 *ptr, __hipsycl_int64 x) {
+  return __hip_atomic_fetch_xor(ptr, x, builtin_memory_order(order),
+                                builtin_memory_scope(scope));
+}
+
+
+
+// TODO: It seems that at least on gfx90a, there are additional unsafe atomic
+// operations that can yield a performance improvement. Should we expose those?
```

Comment:
For reference I think this GROAMCS issue is related. I think we should definitely expose them, however would it be possible to use the `__builtin_amdgcn_ds_atomic_fadd_f32`

Reply:
Usually the set of enabled clang builtins depends on the frontend. It's likely that amdgcn builtins won't work by default with SSCP because the clang frontend does not know that it will run on anything else but the host CPU.

We could just create a simple `__hip/opensycl_sscp_amdgcn_ds_atomic_fadd_f32()` declaration and implement this inside `amdgpu/atomic.cpp` with a simple call to the amdgcn builtin. As long as this function is only invoked when JITing for amdgpu, it should be fine with SSCP.